artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -1567,7 +1566,14 @@ protected void finalize() throws Throwable {
 
    @Override
    public int getConsumerCount() {
-      return consumers.size();
+      // we don't want to count the redistributor, it is an internal transient entry in the consumer list
+      // check for presence before checking consumers to align with use of set()
+      final boolean compensateForPresenceOfRedistributor = redistributorUpdater.get(this) != null;
+      int size = consumers.size();
+      if (size > 0 && compensateForPresenceOfRedistributor) {
+         size--;
+      }
+      return size;
```

Comment:
Sine this isnt done under lock and both the redistributor and consumers list could be changing independently, it seems quite racey? Seems to be used in some important places so I'd wonder if that is a problem.

Given that 'consumers' is actually a custom structure, and updates to its underlying structure are done under lock, it coudl perhaps be made to handle this internally (and e.g add a specific compensating size method)?

Reply:
@gemmellr I like that suggestion. Thanks! I will peek at doing the compensation in size() and have consumers track the redistributor that it contains, it should be a lot cleaner.

Reply:
Just to be clear, I was suggesting adding another method than size. I would expect size to continue to report it as being there (like it does just now) and only things that care about the distinction to call the new one.

Reply:
I think that is all users, the redistributor is managed by the broker, to an observer it does not exist.

Reply:
but I see your point, it may be better in a different method

Reply:
with the latest update, the distinction is now between use of consumers.size when we don't care and getConsumerCount when we only want user consumers.